### DISCUSSION TITLE
ctrl + tab most recent tab switching

### DISCUSSION THREAD

**Discussion by fisher-j** [sumatrapdfreader/sumatrapdf#2750] [Ideas]
I would like to be able to use ctrl + tab to switch tabs in "last viewed" order. Thus, ctrl + tab + tab (tab twice while holding ctrl) would go to the 2nd most recently viewed tab in the queue and ctrl + tab + tab + tab would go to the third and so on. This behavior is common in software with tab interfaces and is useful when you are primarily switching back and forth between two tabs and you do not want to remember tab numbers. 

This is related to #646 where the command palette was suggested as an adequate tab switching mechanism. But sometimes filenames are not all that intuitive and in any case, it is much slower than ctrl + tab.

**Comment by GitHubRulesOK** [comment]
As said before CTRL+Tab in my widows system web /pdf browser (Edge) ALWAYS moves from current tab to the next on the right
this is Identical to what I see in current SumatraPDF behavior.

**Comment by fisher-j** [comment]
You are right, edge browser also does not have this very desirable feature. In that case, it has been left to third party developers to implement [the feature](https://chrome.google.com/webstore/detail/tab-thumbnails-switcher/jpaiaplhepeiilhiegfnknedhjepknng). Of course edge is only one browser. Notably the behavior of alt + tab in Windows is MRU (most recently used) order. MRU tab switching is implemented in firefox and other chromium browsers as well as in popular IDE's . It may not be everyones favorite default behavior, but it is not uncommon and often, users cite this feature as _the_ reason they use a given browser. A setting would allow the user to decide which tab order to use.

**Reply by GitHubRulesOK** [reply]
Well my Firefox 101 must be odd because it works the same windows way too, any order move the tabs or open randomly each time ctrl tab moves to the one on the right ?, You must be setting yours to an enhanced/custom setting. Most of my apps don't have tabs even my acrobat is before tabs introduction, so had to test tracker and guess what however I change opening order or even shuffle tabs, it follows SumatraPDF method too, perhaps its just a windows system thing. my Alt+Tab is commandeered by widows to move from app to app in the same left to right order.

**Comment by tactilis** [comment]
> Well my Firefox 101 must be odd

No it is not odd.

It just depends on the preference you set, which is all @fisher-j is requesting.

I am +1 for this option as well. I find SumatraPDF's current behaviour unhelpful when I ant to switch back and forth between two documents.

![image](https://user-images.githubusercontent.com/4060092/175499842-2568d640-3e5e-4601-ac4d-f8c213ad952b.png)

**Comment by DaPa** [comment]
Added a change in https://github.com/DaPa/sumatrapdf/commit/1664721642b784d17d621c1d324da77b47ffb849 which implements this request via an advanced option **CtrlTabLastViewed** (default _false_). When **CtrlTabLastViewed** = _true_, then Ctrl+Tab behaves as in Firefox, going back to the most recent tab.

**Reply by holmesb** [reply]
Any reason you didn't create a PR @DaPa ? Be great to have this in a Sumatra release.

**Reply by DaPa** [reply]
Sorry for the late answer, I just created a PR in https://github.com/sumatrapdfreader/sumatrapdf/pull/4382. Hope sumatrapdf authors review it carefully and maybe find a better implementation, thanks!

**Comment by holmesb** [comment]
Shame it wasn't merged. SumatraPDF still doesn't have Ctrl+Tab in most recent order. Instead, what was delivered is the same left-to-right order, just with a list of tabs shown.

**Comment by dasebasto** [comment]
In my workflow, I often have to switch between 2 and 4 tabs.

I think that Firefox implemented this "Most Recently Used" (MRU) tab switching the right way, and I love it.

SumatraPDF should add this option: `Ctrl+Tab cycles through tabs in recently used order`

One way to do this: simply order the tab list in the Command Palette in MRU order.

This way, very quick switching becomes possible with Ctrl-Tab, Ctrl-2xTab etc.

**Reply by dasebasto** [reply]
Created a PR implementing Firefox behavior (optional) : https://github.com/sumatrapdfreader/sumatrapdf/pull/5412

**Comment by trACEroam** [comment]
pre-release version 3.7 has resolved this issue for me I think. It has other thoughtful features as well.

**Comment by kjk** [comment]
This is available in pre-release / 3.7 via the `TabsMru` advanced setting.

Set `TabsMru = true` in Settings → Advanced Options… (`SumatraPDF-settings.txt`). Then `Ctrl + Tab` / `Ctrl + Shift + Tab` (Smart Tab Switch) cycle tabs in most recently used order instead of tab-strip order.

Default remains tab-strip order (`TabsMru = false`).

**Comment by holmesb** [comment]
SumatraPDF has just massively levelled-up :-)
